### PR TITLE
Fix DAP ID Issues (primary keys)

### PR DESCRIPTION
**Pull request by quazi-h**
## Why
In our tsv output script, we are generating and appending some artificial primary keys for the DAP tables. This is done so that we can upload the TSVs via the Firecloud api properly since it requires that each TSV have a first column with the label "[entity]:[table_name]_id". The logic for the generated IDs needed a refactor and in addition to that, we have been delivering TSVs without the generated IDs for the benefit of the DAP team.

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1630)

## This PR

- Added an argument to execute branching logic to generate aritificial primary keys as needed for Firecloud uploads.
- Refactored code to generate PKs to address some bugs (ex. 'entity:dog_id_id') brought up by DAP.
- Modified the default logic to maintain existing primary keys and not generate any new ones.
- Updated validation script to accommodate these changes.

### CODE REVIEW

File: hack/convert-output-to-tsv.py [hunk 1]
```diff
@@ -77,29 +71,41 @@ def remove_prefix(text, prefix):
...
-                        log.info(f"Error, 'hs_condition_is_congenital' is not populated in {table_name}")
-                    row[entity_name] = '-'.join([str(component) for component in [row.get('dog_id'), row.get('hs_condition'), congenital_flag]])
-                elif table_name == "environment":
-                    primary_key_fields = ['dog_id', 'address_1_or_2', 'address_month', 'address_year']
-                    row[entity_name] = '-'.join([str(row.get(field)) for field in primary_key_fields])
+                # generate ids for Firecloud compatibility
+                if args.firecloud:
+                    # hles_health_condition: read + copy dog_id and hs_condition and hs_condition_is_congenital
+                    # concatenate and write out as entity_name, keep original PK
+                    if table_name == "hles_health_condition":
```
Comment (aherbst-broad) [inline hack/convert-output-to-tsv.py:78]: 🔧 I think it's time to refactor this a bit, this is a large conditional with many branches that is difficult to reason about. What if we start by pulling lines 78-92 into a separate method that generates a firecloud formatted row?

**Review by aherbst-broad — CHANGES_REQUESTED** [review]
Looking good, I think there's a bit of refactoring to be done to help maintainability.

File: hack/convert-output-to-tsv.py (hunk 1 above)
Comment (ThePletch) [reply hack/convert-output-to-tsv.py:78]: agreed, there's an enormous amount of if statements nested here. i have some ideas for turning this into either a service object or just a function, but that's likely too high-involvement for now.

**Review by ThePletch — CHANGES_REQUESTED** [review]
logic seems sound, but i agree with drew that the sheer size of the functions in this script are beginning to make it difficult to review/reason about. let me know if you want to pair on some refactoring.

File: hack/convert-output-to-tsv.py [hunk 2]
```diff
@@ -36,34 +38,64 @@ def remove_prefix(text, prefix):
 
 gcs = GCSFileSystem()
 
+# create a service object to handle all aspects of generating a primary key
+@dataclass
+class PrimaryKeyGenerator:
```
Comment (aherbst-broad) [inline hack/convert-output-to-tsv.py:43]: 👍 I think this is a better encapsulation of that complexity.

**Review by aherbst-broad — APPROVED** [review]
lgtm

File: hack/convert-output-to-tsv.py [hunk 3]
```diff
@@ -36,34 +38,64 @@ def remove_prefix(text, prefix):
...
+        # most tables should have "dog_id" as a key
+        if self.table_name in {"hles_dog", "hles_cancer_condition", "hles_health_condition", "environment", "cslb"}:
+            self.pk_name = 'dog_id'
+        # owner table is linked to hles_dog via "owner_id"
+        elif self.table_name == 'hles_owner':
+            self.pk_name = 'owner_id'
+        else:
+            raise ValueError(f"Unrecognized table: {self.table_name}")
+
+    # this will return the firecloud entity names
```
Comment (ThePletch) [inline hack/convert-output-to-tsv.py:59]: i think this information is conveyed by the method name, so the comment can be nixed

File: hack/convert-output-to-tsv.py [hunk 4]
```diff
@@ -36,34 +38,64 @@ def remove_prefix(text, prefix):
...
+    def generate_entity_name(self):
+        if self.firecloud:
+            return f"{PRIMARY_KEY_PREFIX}:{self.table_name}_id"
+        return self.pk_name
+
+    # this will return row[entity_name]
```
Comment (ThePletch) [inline hack/convert-output-to-tsv.py:65]: i don't think this comment is correct

File: hack/convert-output-to-tsv.py [hunk 5]
```diff
@@ -36,34 +38,64 @@ def remove_prefix(text, prefix):
...
+        # hles_health_condition: read + copy dog_id and hs_condition and hs_condition_is_congenital
+        # concatenate and write out as entity_name, keep original PK
+        if self.table_name == "hles_health_condition":
+            # code to generate the value of primary key should be pulled out
+            # grab the congenital flag and convert to int
+            congenital_flag = row.get('hs_condition_is_congenital')
+            try:
+                congenital_flag = int(congenital_flag)
+            except TypeError:
+                log.info(f"Error, 'hs_condition_is_congenital' is not populated in {self.table_name}")
```
Comment (ThePletch) [inline hack/convert-output-to-tsv.py:80]: nit: this should be either a `warning` or an `error` instead of an `info`, since it's an error message

**Review by ThePletch — APPROVED** [review]
handful of suggestions about comments and such,  but otherwise i think we're looking good here. thanks for your work on this refactor.

**Pull request merged**